Deprecation and incompatible pointer type warning on new libxml2 versions

```
# github.com/terminalstatic/go-xsd-validate
./libxml2.go: In function ‘cleanup’:
./libxml2.go:116:5: warning: ‘xmlSchemaCleanupTypes’ is deprecated [-Wdeprecated-declarations]
  116 |     xmlSchemaCleanupTypes();
      |     ^~~~~~~~~~~~~~~~~~~~~
In file included from ./libxml2.go:9:
/usr/include/libxml2/libxml/xmlschemastypes.h:37:17: note: declared here
   37 |                 xmlSchemaCleanupTypes           (void);
      |                 ^~~~~~~~~~~~~~~~~~~~~
./libxml2.go: In function ‘parseSchema’:
./libxml2.go:186:5: warning: ‘xmlLineNumbersDefault’ is deprecated [-Wdeprecated-declarations]
  186 |     xmlLineNumbersDefault(1);
      |     ^~~~~~~~~~~~~~~~~~~~~
In file included from /usr/include/libxml2/libxml/tree.h:17,
                 from /usr/include/libxml2/libxml/schemasInternals.h:24,
                 from /usr/include/libxml2/libxml/xmlschemastypes.h:19:
/usr/include/libxml2/libxml/parser.h:967:17: note: declared here
  967 |                 xmlLineNumbersDefault   (int val);
      |                 ^~~~~~~~~~~~~~~~~~~~~
./libxml2.go: In function ‘cParseDoc’:
./libxml2.go:247:5: warning: ‘xmlLineNumbersDefault’ is deprecated [-Wdeprecated-declarations]
  247 |     xmlLineNumbersDefault(1);
      |     ^~~~~~~~~~~~~~~~~~~~~
/usr/include/libxml2/libxml/parser.h:967:17: note: declared here
  967 |                 xmlLineNumbersDefault   (int val);
      |                 ^~~~~~~~~~~~~~~~~~~~~
./libxml2.go: In function ‘cValidate’:
./libxml2.go:300:5: warning: ‘xmlLineNumbersDefault’ is deprecated [-Wdeprecated-declarations]
  300 |     xmlLineNumbersDefault(1);
      |     ^~~~~~~~~~~~~~~~~~~~~
/usr/include/libxml2/libxml/parser.h:967:17: note: declared here
  967 |                 xmlLineNumbersDefault   (int val);
      |                 ^~~~~~~~~~~~~~~~~~~~~
./libxml2.go:328:59: warning: passing argument 2 of ‘xmlSchemaSetValidStructuredErrors’ from incompatible pointer type [-Wincompatible-pointer-types]
  328 |             xmlSchemaSetValidStructuredErrors(schemaCtxt, simpleStructErrorCallback,
      |                                                           ^~~~~~~~~~~~~~~~~~~~~~~~~
      |                                                           |
      |                                                           void (*)(void *, xmlError *) {aka void (*)(void *, struct _xmlError *)}
In file included from /usr/include/libxml2/libxml/xmlschemastypes.h:20:
/usr/include/libxml2/libxml/xmlschemas.h:185:65: note: expected ‘xmlStructuredErrorFunc’ {aka ‘void (*)(void *, const struct _xmlError *)’} but argument is of type ‘void (*)(void *, xmlError *)’ {aka ‘void (*)(void *, struct _xmlError *)’}
  185 |                                          xmlStructuredErrorFunc serror,
      |                                          ~~~~~~~~~~~~~~~~~~~~~~~^~~~~~
./libxml2.go: In function ‘cValidateBuf’:
./libxml2.go:353:5: warning: ‘xmlLineNumbersDefault’ is deprecated [-Wdeprecated-declarations]
  353 |     xmlLineNumbersDefault(1);
      |     ^~~~~~~~~~~~~~~~~~~~~
/usr/include/libxml2/libxml/parser.h:967:17: note: declared here
  967 |                 xmlLineNumbersDefault   (int val);
      |                 ^~~~~~~~~~~~~~~~~~~~~
```

These changes seem to appear in 2.12.0.

I'll submit a PR shortly.

--- a/libxml2.go
+++ b/libxml2.go
@@ -192,7 +192,6 @@ static void simpleStructErrorCallback(
 static struct xsdParserResult parseSchema(
                                           xmlSchemaParserCtxtPtr schemaParserCtxt,
                                           const short int options) {
-    xmlLineNumbersDefault(1);
     bool err = false;
     struct xsdParserResult parserResult;
     errCtx ectx = initErrCtx(1, GO_ERR_INIT);
@@ -253,7 +252,6 @@ static struct xsdParserResult cParseMemSchema(const void* xsd,
 static struct xmlParserResult cParseDoc(const void* goXmlSource,
                                         const int goXmlSourceLen,
                                         const short int options) {
-    xmlLineNumbersDefault(1);
     bool err = false;
     struct xmlParserResult parserResult;
     errCtx ectx = initErrCtx(1, GO_ERR_INIT);
@@ -284,7 +282,7 @@ static struct xmlParserResult cParseDoc(const void* goXmlSource,
                 xmlSetGenericErrorFunc(NULL, noOutputCallback);
             }
 
-            doc = xmlParseMemory(goXmlSource, goXmlSourceLen);
+            doc = xmlReadMemory(goXmlSource, goXmlSourceLen, NULL, NULL, 0);
 
             xmlFreeParserCtxt(xmlParserCtxt);
             if (doc == NULL) {
@@ -306,8 +304,6 @@ static struct xmlParserResult cParseDoc(const void* goXmlSource,
 }
 
 static errArray cValidate(const xmlDocPtr doc, const xmlSchemaPtr schema) {
-    xmlLineNumbersDefault(1);
-
     errArray errArr = initErrArray();
 
     struct simpleXmlError simpleError;
@@ -359,8 +355,6 @@ static errArray cValidateBuf(const void* goXmlSource,
                              const int goXmlSourceLen,
                              const short int xmlParserOptions,
                              const xmlSchemaPtr schema) {
-    xmlLineNumbersDefault(1);
-
     errArray errArr = initErrArray();
 
     struct simpleXmlError simpleError;

--- a/libxml2.go
+++ b/libxml2.go
@@ -113,7 +113,9 @@ static void init() {
 }
 
 static void cleanup() {
+#if LIBXML_VERSION < 21000
     xmlSchemaCleanupTypes();
+#endif
     xmlCleanupParser();
 }
 

--- a/libxml2.go
+++ b/libxml2.go
@@ -141,7 +141,14 @@ static void genErrorCallback(void* ctx, const char* message, ...) {
     free(newLine);
 }
 
-static void simpleStructErrorCallback(void* ctx, xmlErrorPtr p) {
+static void simpleStructErrorCallback(
+    void* ctx,
+#if LIBXML_VERSION >= 21200
+    const xmlError *p
+#else
+    xmlErrorPtr p
+#endif
+) {
     errArray* sErrArr = ctx;
 
     struct simpleXmlError sErr;

